app/database/build_store.py

```diff
@@ -0,0 +1,439 @@
+from collections import OrderedDict
+import json
+from queue import Queue
+from typing import List, Optional
+from sqlalchemy import func
+
+from app.database.connection import Connection
+from app.database.schema import AtomsSchema, BuildSchema, FailedArtifactDirectoriesSchema, FailedSubjobAtomPairsSchema, SubjobsSchema
+from app.common.build_artifact import BuildArtifact
+from app.master.atom import Atom
+from app.master.build import Build
+from app.master.build_request import BuildRequest
+from app.master.build_fsm import BuildState
+from app.master.subjob import Subjob
+from app.util.exceptions import ItemNotFoundError
+from app.util.log import get_logger
+
+
+# pylint: disable=protected-access
+class BuildStore:
+    """
+    Build storage service that stores and handles all builds.
+    """
+    _logger = get_logger(__name__)
+    _cached_builds_by_id = OrderedDict()
+
+    @classmethod
+    def get(cls, build_id: int, allow_incompleted_builds=False) -> Optional[Build]:
+        """
+        Returns a build by id.
+        :param build_id: The id for the build whose status we are getting
+        """
+        build = cls._cached_builds_by_id.get(build_id)
+        if build is None:
+            cls._logger.info('Requested build (id: {}) was not found in cache, checking database.'.format(build_id))
+            build = cls._reconstruct_build(build_id, allow_incompleted_builds=allow_incompleted_builds)
+            if build is not None:
+                cls._cached_builds_by_id[build_id] = build
+                cls._logger.notice('Build (id: {}) was added to cache.'.format(build_id))
```

Comment:
Some of these logging messages can get really noising, I wouldn't be opposed to removing them 👍